polygon/sync/canonical_chain_builder.go

```diff
@@ -0,0 +1,246 @@
+	headers := impl.Headers()
+	if len(headers) == 0 {
+		return nil
+	}
+	if headers[0].Number.Uint64() > start {
+		return nil
+	}
+	if headers[len(headers)-1].Number.Uint64() < start+count-1 {
+		return nil
+	}
+
+	offset := start - headers[0].Number.Uint64()
+	return headers[offset : offset+count]
+}
+
+func (impl *canonicalChainBuilderImpl) Prune(newRootNum uint64) error {
+	if (newRootNum < impl.root.header.Number.Uint64()) || (newRootNum > impl.Tip().Number.Uint64()) {
+		return errors.New("canonicalChainBuilderImpl.Prune: newRootNum outside of the canonical chain")
+	}
+
+	newRoot := impl.tip
+	for newRoot.header.Number.Uint64() > newRootNum {
+		newRoot = newRoot.parent
+	}
+
+	impl.root = newRoot
+	return nil
+}
+
+func mapSlice[T any, U any](s []T, mapFunc func(T) U) []U {
+	out := make([]U, 0, len(s))
+	for _, x := range s {
+		out = append(out, mapFunc(x))
+	}
+	return out
+}
+
+func (impl *canonicalChainBuilderImpl) updateTipIfNeeded(tipCandidate *forkTreeNode) {
+	if tipCandidate.totalDifficulty > impl.tip.totalDifficulty {
+		impl.tip = tipCandidate
+	} else if tipCandidate.totalDifficulty == impl.tip.totalDifficulty {
+		// TODO: is it possible? which one is selected?
+	}
+}
+
+func (impl *canonicalChainBuilderImpl) Connect(headers []*types.Header) error {
+	if (len(headers) > 0) && (headers[0].Number != nil) && (headers[0].Number.Cmp(impl.root.header.Number) == 0) {
+		headers = headers[1:]
+	}
+	if len(headers) == 0 {
+		return nil
+	}
+
+	parent := impl.nodeByHash(headers[0].ParentHash)
+	if parent == nil {
+		return errors.New("canonicalChainBuilderImpl.Connect: can't connect headers")
+	}
+
+	headersHashes := mapSlice(headers, func(header *types.Header) libcommon.Hash {
+		return header.Hash()
+	})
+
+	// check if headers are linked by ParentHash
+	for i, header := range headers[1:] {
+		if header.ParentHash != headersHashes[i] {
+			return errors.New("canonicalChainBuilderImpl.Connect: invalid headers slice ParentHash")
+		}
+	}
+
+	// skip existing matching nodes until a new header is found
+	for len(headers) > 0 {
+		var matchingNode *forkTreeNode
+		for _, c := range parent.children {
+			if c.headerHash == headersHashes[0] {
+				matchingNode = c
+				break
+			}
+		}
+		if matchingNode != nil {
+			parent = matchingNode
+			headers = headers[1:]
+			headersHashes = headersHashes[1:]
+		} else {
+			break
+		}
+	}
+
+	// if all headers are already inserted
+	if len(headers) == 0 {
+		return nil
+	}
+
+	// attach nodes for the new headers
+	for i, header := range headers {
+		if (header.Number == nil) && (header.Number.Uint64() != parent.header.Number.Uint64()+1) {
+			return errors.New("canonicalChainBuilderImpl.Connect: invalid header.Number")
+		}
+
+		// TODO: validate using CalcProducerDelay
+		if header.Time <= parent.header.Time {
+			return errors.New("canonicalChainBuilderImpl.Connect: invalid header.Time")
+		}
+
+		if err := bor.ValidateHeaderExtraField(header.Extra); err != nil {
```

Comment:
Have a look at [checkHeaderExtraData](https://github.com/ledgerwatch/erigon/blob/devel/eth/stagedsync/stage_bor_heimdall.go#L356) - when the header is the last of a sprint we do more validations - check the span selected proposers match the validator bytes in the header extra data. 

Maybe add a todo if not adding as part of this PR?

On another note it may be a good idea if we write down a list of all validations we should be doing and double check the code later on that we have them all?

Reply:
:+1: Yes, I planned to do more validations in the next PRs and wanted to discuss some of them.